scheduler/feasible.go

```diff
@@ -237,10 +238,111 @@ func (iter *ProposedAllocConstraintIterator) satisfiesDistinctHosts(option *stru
 	return true
 }
 
-func (iter *ProposedAllocConstraintIterator) Reset() {
+func (iter *DistinctHostsIterator) Reset() {
 	iter.source.Reset()
 }
 
+// DistinctPropertyIterator is a FeasibleIterator which returns nodes that pass the
+// distinct_property constraint. The constraint ensures that multiple allocations
+// do not use the same value of the given property.
+type DistinctPropertyIterator struct {
+	ctx    Context
+	source FeasibleIterator
+	tg     *structs.TaskGroup
+	job    *structs.Job
+
+	hasDistinctPropertyConstraints bool
+	jobPropertySets                []*propertySet
+	groupPropertySets              map[string][]*propertySet
+}
+
+// NewDistinctPropertyIterator creates a DistinctPropertyIterator from a source.
+func NewDistinctPropertyIterator(ctx Context, source FeasibleIterator) *DistinctPropertyIterator {
+	return &DistinctPropertyIterator{
+		ctx:               ctx,
+		source:            source,
+		groupPropertySets: make(map[string][]*propertySet),
+	}
+}
+
+func (iter *DistinctPropertyIterator) SetTaskGroup(tg *structs.TaskGroup) {
+	iter.tg = tg
+
+	// Check if there is a distinct property
+	iter.hasDistinctPropertyConstraints = len(iter.jobPropertySets) != 0 || len(iter.groupPropertySets[tg.Name]) != 0
+}
+
+func (iter *DistinctPropertyIterator) SetJob(job *structs.Job) {
+	iter.job = job
+
+	// Build the property sets
+	for _, c := range job.Constraints {
+		if c.Operand != structs.ConstraintDistinctProperty {
+			continue
+		}
+
+		pset := NewPropertySet(iter.ctx, job)
+		pset.SetJobConstraint(c)
+		iter.jobPropertySets = append(iter.jobPropertySets, pset)
+	}
+
+	for _, tg := range job.TaskGroups {
+		for _, c := range tg.Constraints {
+			if c.Operand != structs.ConstraintDistinctProperty {
+				continue
+			}
+
+			pset := NewPropertySet(iter.ctx, job)
+			pset.SetTGConstraint(c, tg.Name)
+			iter.groupPropertySets[tg.Name] = append(iter.groupPropertySets[tg.Name], pset)
+		}
+	}
+}
+
+func (iter *DistinctPropertyIterator) Next() *structs.Node {
+OUTER:
+	for {
+		// Get the next option from the source
+		option := iter.source.Next()
+
+		// Hot path if there is nothing to check
+		if option == nil || !iter.hasDistinctPropertyConstraints {
+			return option
+		}
+
+		// Check if the constraints are met
+		for _, ps := range iter.jobPropertySets {
```

Comment:
You can probably just have a helper method for `satisfiesProperties()` which takes a list, and call it once for `jobPropertySets` and `groupPropertySets` to DRY